### PR TITLE
route: improve route entry for matcher

### PR DESCRIPTION
**Pull request by tony612**
Commit Message: route: improve route entry for matcher

* support dynamic clusters for matcher
* run route callback after route match for matcher

Additional Description: 

There are 2 problems in RouteAction and RouteList of matcher:
1. After route matches, the route in the config will be returned directly. But we should return the result of `matches`. Otherwise, the dynamic cluster(like cluster header) won't work.
2. They doesn't handle the route callback logic.
(I hit the first problem, but found the second when I fixes the first)

Without the fix, the tests will fail:
```
[ RUN      ] RouteMatcherTest.TestMatchTreeDynamicCluster
external/envoy/test/common/router/config_impl_test.cc:1430: Failure
Expected equality of these values:
  "www2"
  route->clusterName()
    Which is: ""

[ RUN      ] RouteMatcherTest.TestRouteListDynamicCluster
external/envoy/test/common/router/config_impl_test.cc:1585: Failure
Expected equality of these values:
  "www2"
  route->clusterName()
    Which is: ""

[ RUN      ] RouteMatchOverrideTest.RouteListVerifyRouteOverrideStops
external/envoy/test/common/router/config_impl_test.cc:10748: Failure
Expected equality of these values:
  accepted_route->routeEntry()->clusterName()
    Which is: "foo_bar"
  "foo"
```

Risk Level: Low
Testing: unit tests
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/router/config_impl.cc [hunk 1]
```diff
@@ -1924,10 +1954,6 @@ RouteConstSharedPtr VirtualHostImpl::getRouteFromEntries(const RouteCallback& cb
         if (match_status == RouteMatchStatus::Accept) {
           return route_entry;
         }
-        if (match_status == RouteMatchStatus::Continue &&
-            eval_status == RouteEvalStatus::NoMoreRoutes) {
-          return nullptr;
-        }
```
Comment (tony612) [inline source/common/router/config_impl.cc:1930]: This is a little refactoring. When match_status is Continue and eval_status is NoMoreRoutes, returning nullptr is the same as continuing the loop. After the refactoring, the code in matcher will be simpler.

File: source/common/router/config_impl.cc [hunk 2]
```diff
@@ -1879,20 +1879,50 @@ RouteConstSharedPtr VirtualHostImpl::getRouteFromEntries(const RouteCallback& cb
...
+        RouteConstSharedPtr route_entry =
+            route_action.route()->matches(headers, stream_info, random_value);
+        if (nullptr == route_entry) {
+          ENVOY_LOG(debug, "route was resolved but final route did not match incoming request");
+          return nullptr;
         }
 
-        ENVOY_LOG(debug, "route was resolved but final route did not match incoming request");
-        return nullptr;
+        if (cb) {
```
Comment (silverstar194) [inline source/common/router/config_impl.cc:1893]: Can you please unnest here?

File: source/common/router/config_impl.cc [hunk 3]
```diff
@@ -1879,20 +1879,50 @@ RouteConstSharedPtr VirtualHostImpl::getRouteFromEntries(const RouteCallback& cb
...
         const RouteMatchAction& route_action = result->getTyped<RouteMatchAction>();
+        if (!headers.Path() && !route_action.route()->supportsPathlessHeaders()) {
+          return nullptr;
+        }
 
-        if (route_action.route()->matches(headers, stream_info, random_value)) {
-          return route_action.route();
+        RouteConstSharedPtr route_entry =
+            route_action.route()->matches(headers, stream_info, random_value);
+        if (nullptr == route_entry) {
```
Comment (silverstar194) [inline source/common/router/config_impl.cc:1888]: route_entry == nullptr

File: source/common/router/config_impl.cc [hunk 4]
```diff
@@ -1879,20 +1879,50 @@ RouteConstSharedPtr VirtualHostImpl::getRouteFromEntries(const RouteCallback& cb
...
-        for (const auto& route : action.routes()) {
-          if (route->matches(headers, stream_info, random_value)) {
-            return route;
+        for (auto route = action.routes().begin(); route != action.routes().end(); ++route) {
+          if (!headers.Path() && !(*route)->supportsPathlessHeaders()) {
+            continue;
           }
+
+          RouteConstSharedPtr route_entry = (*route)->matches(headers, stream_info, random_value);
+          if (nullptr == route_entry) {
```
Comment (silverstar194) [inline source/common/router/config_impl.cc:1910]: switch order

File: source/common/router/config_impl.cc [hunk 5]
```diff
@@ -1879,20 +1879,50 @@ RouteConstSharedPtr VirtualHostImpl::getRouteFromEntries(const RouteCallback& cb
...
+                                              ? RouteEvalStatus::NoMoreRoutes
+                                              : RouteEvalStatus::HasMoreRoutes;
+            RouteMatchStatus match_status = cb(route_entry, eval_status);
+            if (match_status == RouteMatchStatus::Accept) {
+              return route_entry;
+            }
+            continue;
+          }
+
+          return route_entry;
```
[the remark this answers is not in the document]
Comment (tyxia) [reply source/common/router/config_impl.cc:1925]: Besides, I suggest refactor the code like below to reduce nested if and redundant continue, which improves the readability. 
(from line 1914)

```
if (cb == nullptr) {
        return route_entry;
}
RouteEvalStatus eval_status = (std::next(route) == routes_.end())
                                  ? RouteEvalStatus::NoMoreRoutes
                                  : RouteEvalStatus::HasMoreRoutes;

RouteMatchStatus match_status = cb(route_entry, eval_status);
if (match_status == RouteMatchStatus::Accept) {
  return route_entry;
}
if (match_status == RouteMatchStatus::Continue &&
    eval_status == RouteEvalStatus::NoMoreRoutes) {
  return nullptr;
}

```

File: source/common/router/config_impl.cc [hunk 6]
```diff
@@ -1879,20 +1879,50 @@ RouteConstSharedPtr VirtualHostImpl::getRouteFromEntries(const RouteCallback& cb
...
-        ENVOY_LOG(debug, "route was resolved but final route did not match incoming request");
-        return nullptr;
+        if (cb) {
+          if (cb(route_entry, RouteEvalStatus::NoMoreRoutes) == RouteMatchStatus::Accept) {
+            return route_entry;
+          }
+          return nullptr;
+        }
+
+        return route_entry;
```
Comment (tyxia) [inline source/common/router/config_impl.cc:1900]: The code block here (from 1882 to 1900) can also  be replaced with helper function I suggested below.  Because it is just the case where route vector has 1 element (i.e., {route_action.route()})

Basically, we replace these three big code blocks to three  one-line func calls to the helper function.

**Review by tyxia — COMMENTED** [review]
Thanks for working on this.

Left some comments about code refactor to make the code nicer

File: source/common/router/config_impl.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (tony612) [reply source/common/router/config_impl.cc]: Done. Is the current code good for you? btw, I'm not very familiar with c++, just wonder if it will be better to change `getRouteFromRoutes` to an inline function in this case?

**Review by silverstar194 — DISMISSED** [review]

**Comment by tony612** [comment]
Thanks! Conflicts resolved.

**Review by tyxia — COMMENTED** [review]
Mostly LGTM. Thanks!

**Review by tyxia — DISMISSED** [review]
LGTM, Thanks!

File: test/common/router/config_impl_test.cc [hunk 7]
```diff
@@ -1490,7 +1525,7 @@ TEST_F(RouteMatcherTest, TestRouteList) {
   {
     Http::TestRequestHeaderMapImpl headers =
         genHeaders("lyft.com", "/new_endpoint/foo/match_header", "GET");
-    headers.setByKey("x-match-header", "matched");
+    headers.addCopy("x-match-header", "matched");
```
Comment (wbpcode) [inline test/common/router/config_impl_test.cc:1528]: Is this change necessary?

File: source/common/router/config_impl.h [hunk 8]
```diff
@@ -321,6 +321,11 @@ class VirtualHostImpl : public VirtualHost, Logger::Loggable<Logger::Id::router>
                              scope.scopeFromStatName(stat_names.other_), stat_names) {}
   };
 
+  RouteConstSharedPtr
+  getRouteFromRoutes(const RouteCallback& cb, const Http::RequestHeaderMap& headers,
+                     const StreamInfo::StreamInfo& stream_info, uint64_t random_value,
+                     std::vector<RouteEntryImplBaseConstSharedPtr> routes) const;
```
Comment (wbpcode) [inline source/common/router/config_impl.h:327]: So, I will always copy the whole route list? May be `Span` would be better here to avoid any unnecessary stack allocation.

**Review by wbpcode — CHANGES_REQUESTED** [review]
Thanks for your contribution and some comments are added. 😸

File: test/common/router/config_impl_test.cc (hunk 7 above)
Comment (tony612) [reply test/common/router/config_impl_test.cc:1528]: Not necessary. When I set the header in TestMatchTreeDynamicCluster, I found `setByKey` may cause the lifetime problem and I noticed `setByKey` is barely used in Envoy other tests. So I also changed my previous code.

Comment (wbpcode) [reply test/common/router/config_impl_test.cc:1528]: yeah, `setByKey` is api from `TraceContext`. If you want keep this change, then `setCopy` would be better here to keep the semantic.

Comment (tony612) [reply test/common/router/config_impl_test.cc]: Got it. Done.

File: source/common/router/config_impl.h (hunk 8 above)
Comment (tony612) [reply source/common/router/config_impl.h]: Done. Looks like `Span` can also avoid copying, so changing vector to `Span` is enough, right?

File: source/common/router/config_impl.h [hunk 9]
```diff
@@ -324,7 +324,7 @@ class VirtualHostImpl : public VirtualHost, Logger::Loggable<Logger::Id::router>
   RouteConstSharedPtr
   getRouteFromRoutes(const RouteCallback& cb, const Http::RequestHeaderMap& headers,
                      const StreamInfo::StreamInfo& stream_info, uint64_t random_value,
-                     std::vector<RouteEntryImplBaseConstSharedPtr> routes) const;
+                     absl::Span<const RouteEntryImplBaseConstSharedPtr> routes) const;
```
Comment (wbpcode) [inline source/common/router/config_impl.h:327]: is this `const` necessary?

Comment (tony612) [reply source/common/router/config_impl.h:327]: Yes. Otherwise, the vector can't be converted to absl::Span:
```
source/common/router/config_impl.cc:1920:75: error: no matching constructor for initialization of 'absl::Span<RouteEntryImplBaseConstSharedPtr>' (aka 'Span<shared_ptr<const Envoy::Router::RouteEntryImplBase>>')

source/common/router/config_impl.cc:1925:75: error: no viable conversion from 'const std::vector<RouteEntryImplBaseConstSharedPtr>' (aka 'const vector<shared_ptr<const Envoy::Router::RouteEntryImplBase>>') to 'absl::Span<RouteEntryImplBaseConstSharedPtr>' (aka 'Span<shared_ptr<const Envoy::Router::RouteEntryImplBase>>')

source/common/router/config_impl.cc:1936:69: error: no viable conversion from 'const std::vector<RouteEntryImplBaseConstSharedPtr>' (aka 'const vector<shared_ptr<const Envoy::Router::RouteEntryImplBase>>') to 'absl::Span<RouteEntryImplBaseConstSharedPtr>' (aka 'Span<shared_ptr<const Envoy::Router::RouteEntryImplBase>>')
```

Comment (wbpcode) [reply source/common/router/config_impl.h:327]: Hmmm, a little weird. The `Span` should be const by default just like the `absl::string_view`. Anyway, it isn't big matter.

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks. But this patch updates very core code. So, I will defer the merge to a senior maintainer.
